src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -1231,6 +1256,61 @@ ep_buffer_manager_write_all_buffers_to_file_v4 (
+					!ep_rt_sequence_point_list_iterator_end(&buffer_manager->sequence_points, &sequence_point_list_iterator);
+					ep_rt_sequence_point_list_iterator_next(&sequence_point_list_iterator)) {
+
+					sequence_point = ep_rt_sequence_point_list_iterator_value(&sequence_point_list_iterator);
+
+					// foreach (session_state in session_states_to_delete)
+					for (ep_rt_thread_session_state_array_iterator_t thread_session_state_array_iterator = ep_rt_thread_session_state_array_iterator_begin(&session_states_to_delete);
+						!ep_rt_thread_session_state_array_iterator_end(&session_states_to_delete, &thread_session_state_array_iterator);
+						ep_rt_thread_session_state_array_iterator_next(&thread_session_state_array_iterator)) {
+
+						EventPipeThreadSessionState * thread_session_state = ep_rt_thread_session_state_array_iterator_value(&thread_session_state_array_iterator);
+						uint32_t thread_sequence_number = 0;
+						bool exists = ep_rt_thread_sequence_number_map_lookup (ep_sequence_point_get_thread_sequence_numbers_cref (sequence_point), thread_session_state, &thread_sequence_number);
+						if (exists) {
+							ep_rt_thread_sequence_number_map_remove (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), thread_session_state);
+							// every entry of this map was holding an extra ref to the thread (see: ep-event-instance.{h|c})
+							ep_thread_release (ep_thread_session_state_get_thread (thread_session_state));
+						}
+					}
+				}
+			}
+
+		EP_SPIN_LOCK_EXIT (&buffer_manager->rt_lock, section4)
+	}
+
+	// foreach (session_state in session_states_to_delete)
+	for (ep_rt_thread_session_state_array_iterator_t thread_session_state_array_iterator = ep_rt_thread_session_state_array_iterator_begin(&session_states_to_delete);
+			!ep_rt_thread_session_state_array_iterator_end(&session_states_to_delete, &thread_session_state_array_iterator);
+			ep_rt_thread_session_state_array_iterator_next(&thread_session_state_array_iterator)) {
+
+		EventPipeThreadSessionState * thread_session_state = ep_rt_thread_session_state_array_iterator_value(&thread_session_state_array_iterator);
+		EP_ASSERT (thread_session_state != NULL);
+		// This may be the last reference to a given EventPipeThread, so make a ref to keep it around till we're done
+		EventPipeThreadHolder thread_holder;
+		if (ep_thread_holder_init (&thread_holder, ep_thread_session_state_get_thread (thread_session_state))) {
+
+			// briefly hold the thread lock to check it's state
+			ep_rt_spin_lock_handle_t *thread_lock = ep_thread_get_rt_lock_ref (ep_thread_session_state_get_thread (thread_session_state));
+			EP_SPIN_LOCK_ENTER (thread_lock, section5)
+				// if the thread is unregistered delete the session state from it
+				if (ep_rt_volatile_load_uint32_t_without_barrier(&(ep_thread_holder_get_thread (&thread_holder))->unregistered) > 0) {
```

Comment:
You should access unregistered filed through a getter since it is declared as part of ep-thread.c and not directly accessible to ep-buffer-manager (if compiled without inline getter/setter), so it is a "private" field.